thorlcr/activities/keyedjoin/thkeyedjoinslave.cpp

```diff
@@ -2795,6 +3267,329 @@ class CKeyedJoinSlave : public CSlaveActivity, implements IJoinProcessor
+                offset_t offset = 0;
+                if (orderedParts)
+                {
+                    spillMarkers.resize(orderedParts);
+                    size32_t len = sizeof(CJoinGroup::SpillMarker)*orderedParts;
+                    iFileIO->read(0, len, &spillMarkers[0]);
+                    currentPart = 0;
+                    while (true)
+                    {
+                        if (spillMarkers[currentPart].size)
+                        {
+                            offset = ((currentPart+1) * sizeof(CJoinGroup::SpillMarker)) + spillMarkers[currentPart].skip;
+                            break;
+                        }
+                        ++currentPart;
+                        assertex(currentPart != orderedParts); // Should never get here, implies nothing written
+                    }
+                    remaining = spillMarkers[currentPart].size;
+                }
+                strm.setown(createFileSerialStream(iFileIO, offset));
+                prefetchBuffer.setStream(strm);
+            }
+            virtual const void *getRow(offset_t &fpos) override
+            {
+                if (eos)
+                    return nullptr;
+
+                RtlDynamicRowBuilder rowBuilder(*allocator);
+                size32_t size = deserializer->deserialize(rowBuilder, prefetchBuffer);
+                const void *row = rowBuilder.finalizeRowClear(size);
+                prefetchBuffer.read(sizeof(fpos), &fpos);
+                prefetchBuffer.finishedRow();
+
+                /* NB: eos is marked by hitting terminator marker if ordered
+                 * If not, eos is at end of stream
+                 */
+                if (orderedParts)
+                {
+                    remaining -= size + sizeof(fpos);
+                    if (0 == remaining)
+                    {
+                        CJoinGroup::SpillMarker marker;
+                        const void *ptr = prefetchBuffer.peek(sizeof(CJoinGroup::SpillMarker));
+                        memcpy(&marker, ptr, sizeof(CJoinGroup::SpillMarker));
+                        prefetchBuffer.skip(sizeof(CJoinGroup::SpillMarker));
+                        remaining = marker.size;
+                        if (0 == remaining)
+                        {
+                            while (true)
+                            {
+                                ++currentPart;
+                                if (currentPart == orderedParts)
+                                {
+                                    eos = true;
+                                    break;
+                                }
+                                if (spillMarkers[currentPart].size)
+                                {
+                                    offset_t offset = ((currentPart+1) * sizeof(CJoinGroup::SpillMarker)) + spillMarkers[currentPart].skip;
+                                    prefetchBuffer.reset(offset);
```

Comment:
>As it stands I think this will throw away the decompressed buffer

there is no decompression involved when writing like this (when writing incomplete groups), because compressed writer doesn't support random seek and the way it needs to patch the markers require it.

Reply:
>Am I correct that there is a separate file created for each join group?

Yes.

Reply:
>I think it might be common for offset to match the current position (as long as there have not been multiple spills)

yes it might, so I could change it to:
```
                                    if (prefetchBuffer.tell() != offset)
                                        prefetchBuffer.reset(offset);

```

but perhaps that test should to inside CContiguousRowBuffer::reset() instead.